### PR TITLE
Gutenberg 3.2

### PR DESCRIPTION
**Pull request by apardyl**
* Add multiple printer support with printer selection,
* Extend the permission system to support per printer restrictions,
* Replace old printing form with a SPA with PWA capabilities,
* Minor fixes.
* Monitor job status with lpstat.

### CODE REVIEW

File: ipp/views.py [hunk 1]
```diff
@@ -43,5 +44,8 @@ def post(self, request: HttpRequest, token, rel_path, *args, **kwargs):
                 return res
             else:
                 return HttpResponse(b'Forbidden', status=403, content_type='text/plain')
-        service = IppService(user, request.is_secure(), basic_auth)
+        printer = Printer.get_printer_for_user(user, printer_id)
+        if not printer:
+            return HttpResponse(b'Not found', status=404, content_type='text/plain')
```
Comment (boorus) [inline ipp/views.py:49]: Consider raising `django.http.Http404`.

File: printing/converter.py [hunk 2]
```diff
@@ -31,8 +31,8 @@ def is_available(cls):
 
 class SandboxConverter(Converter, ABC):
     def run_in_sandbox(self, command: List[str]):
-        subprocess.check_call(
-            [SANDBOX_PATH, self.work_dir] + command)
+        subprocess.check_output(
```
Comment (boorus) [inline printing/converter.py:34]: What's the reason behind switching to `check_output` here?

File: printing/printing.py [hunk 3]
```diff
@@ -1,61 +1,178 @@
-# Original script by zielmicha (https://github.com/zielmicha)
-
+import logging
```
Comment (boorus) [inline printing/printing.py:1]: Why is this import in a separate grup? Please consider merging them together.

File: printing/printing.py [hunk 4]
```diff
@@ -1,61 +1,178 @@
...
 
-def generate_hp500_options(properties: PrintingProperties):
-    options = []
-    options += ['-n', str(properties.copies)]
-    options += ['-o', 'HPColorAsGray={}'.format(not properties.color)]
+    def __init__(self) -> None:
+        self.backend_name = self.__class__.__name__
+
+    @abstractmethod
+    def check_status(self, job: PrintJob, backend_job_id: Any) -> bool:
```
Comment (boorus) [inline printing/printing.py:42]: Please consider adding a docstring for the methods here.

File: printing/printing.py [hunk 5]
```diff
@@ -1,61 +1,178 @@
...
 
+logger = logging.getLogger('gutenberg.worker')
+
+
+def handle_cancellation(job: PrintJob, handler: Optional[Callable[[], None]] = None):
+    # We allow a low possibility of a race condition here as the impact would be negligible
+    # (ie. ignored request) and the probability is low.
+    job.refresh_from_db()
+    if job.status == JobStatus.CANCELING:
+        logger.info("Cancelling job {}".format(job))
```
Comment (boorus) [inline printing/printing.py:27]: Please consider switching to American English spelling: `Canceling`.

File: printing/printing.py [hunk 6]
```diff
@@ -1,61 +1,178 @@
...
+    def print(self, job: PrintJob, file_path: str):
+        logger.info("Printing job {} via {}".format(job, self.backend_name))
+        backend_job_id = self.submit_job(job, file_path)
+        cnt = 0
+        while self.check_status(job, backend_job_id):
+            logger.info("Job {} is still printing via {}".format(job, self.backend_name))
+            handle_cancellation(job, lambda: self.cancel_job(job, backend_job_id))
+            time.sleep(1)
+            cnt += 1
+            if cnt > 120:
```
Comment (boorus) [inline printing/printing.py:62]: Please consider extracting this magic constant.

File: printing/printing.py [hunk 7]
```diff
@@ -1,61 +1,178 @@
...
-        TwoSidedPrinting.ONE_SIDED: 'None',
-        TwoSidedPrinting.TWO_SIDED_LONG_EDGE: 'DuplexNoTumble',
-        TwoSidedPrinting.TWO_SIDED_SHORT_EDGE: 'DuplexTumble',
-    }[properties.two_sides]
-    options += ['-o', 'Duplex={}'.format(two_sided_opt)]
-    options += ['-o', 'fit-to-page']
-    return options
 
+class LocalCupsPrinter(PrinterBackend):
+    CUPS_JOB_REQUEST_REGEX = r'^request id is ([^ ]+) '
```
Comment (boorus) [inline printing/printing.py:72]: Is this locale-resistant?

File: printing/printing.py [hunk 8]
```diff
@@ -1,61 +1,178 @@
...
-            subprocess.check_call(
-                ['lp', out] + ['-d', settings.PRINTER_NAME] +
-                generate_options(job.properties))
-            job.status = JobStatus.COMPLETED
-            job.finished = timezone.now()
-            job.save()
+            backend = {
+                PrinterType.DISABLED: DisabledPrinter,
+                PrinterType.LOCAL_CUPS: LocalCupsPrinter,
+            }.get(job.printer.printer_type, PrinterType.DISABLED)()
```
Comment (boorus) [inline printing/printing.py:167]: Did you mean `DisabledPrinter` as a default value?

File: printing/postprocess.py [hunk 9]
```diff
@@ -1,21 +1,36 @@
...
+    subprocess.check_output([SANDBOX_PATH, work_dir, 'gs', '-sDEVICE=pdfwrite', '-dNOPAUSE',
+                             '-dBATCH', '-dSAFER', '-dCompatibilityLevel=1.4',
+                             '-sOutputFile=' + out, input_file], stderr=subprocess.STDOUT)
 
+    properties = job.properties
     if properties.pages_to_print:
         new_out = os.path.join(work_dir, 'post.pdf')
-        pages = properties.pages_to_print.replace(',', ' ').split()
-        subprocess.check_call([SANDBOX_PATH, work_dir, 'pdftk', out, 'cat', *pages, 'output', new_out])
+        pages = properties.pages_to_print.split(',')
```
Comment (boorus) [inline printing/postprocess.py:26]: Is `properties.pages_to_print` somehow normalized here? Otherwise we'll handle arguments such as `1 2,4` incorrectly.

**Review by boorus — CHANGES_REQUESTED** [review]

File: control/models.py [hunk 10]
```diff
@@ -26,9 +30,81 @@ class TwoSidedPrinting(models.TextChoices):
...
+
+    def __str__(self):
+        return '{} - {}'.format(self.printer, self.group)
+
+    class Meta:
+        unique_together = ('printer', 'group')
+
+    @staticmethod
+    def get_permissions(user: User, printer: Printer) -> Set[PrinterPermissionsEnum]:
+        if False and user.is_superuser:
```
[the remark this answers is not in the document]
Comment (apardyl) [reply control/models.py:92]: Good catch, forgot to remove test code.

File: printing/postprocess.py (hunk 9 above)
Comment (apardyl) [reply printing/postprocess.py:26]: I forgot to migrate server-side validation from the old form-based UI. Fixed, not it is required to be in a specific format.

**Comment by apardyl** [comment]
LGTM'ed over messenger by @m4tx

**Pull request merged**